### PR TITLE
load databases without accessing classes

### PR DESCRIPTION
**Pull request by mringler**
This MR changes the way tables are registered during initialization to improve performance. Instead of building the internal table lookup structure by accessing static class properties of the table classes, the table lookup is loaded directly from the init file.

#### Background

During initialization, available database tables are registered in Propel. Currently this is done through the qualified class name (i.e. `\Propel\Tests\BookstoreSchemas\Map\BookTableMap`), which is handed to a `DatabaseMap`, which then adds it to its table lookup maps. There are two kinds of lookup, access by name (schema and table name, i.e. `bookstore_schemas.book`) or by PHP name (either set in schema.xml or generated from name, i.e. `\Book`).
Currently, the lookup maps are created by accessing static properties of the supplied class. It was reported that this leads to a measurable slowdown when working with lots of tables (300+), see [here](https://github.com/propelorm/Propel2/pull/1909#issuecomment-1292112844).

#### Proposed Solution
Instead of handing over only the table class name to the DatabaseMap, we can also hand over the name and PHP name, making the static access obsolete. Even simpler, instead of reading the raw table data from the init file (usually `loadDatabase.php`) and then building up the lookup structure every time, we can just write and then retrieve that lookup structure from the init file, and simply load it into the DatabaseMap as it is.
#1909 does something similar, and it reportedly leads to a speed up.

#### Changes
* The `DatabaseMap` class gets a method to dump its lookup maps and a method to load them again from dump.
* The `TableMapLoaderScriptBuilder` class (which generates the `loadDatabase.php` file) does not just write an array of table class names, but registers the tables in a local DatabaseMap, and then writes a dump to the init file.
* During initialization, the dump is loaded back into the DatabaseMaps (through `StandardServiceContainer`).

The changes do not impact BC, even after the update, an old init file can still be used.

#### Result
On my machine, with PHP 8.1 and a project with 40 tables, initialization through static access takes around 0.8 ms (+- 0.3ms), while initialization through dumps takes around 0.25 ms (+- 0.1ms). I am actually surprised how clearly and distinct the difference between the two approaches is. I guess it comes from the time it takes PHP to load the classes, but there is also the repeated calls to a subroutine and writing to the lookup arrays individually. I have tested it through the php dev server, not on an actual webserver, which might give different results.
Of course we are talking about less than a thousand of a second, so this will not make a difference to most users. But it won't hurt either.

I don't have a project with more tables available, but maybe @profuel can supply some measurements for a larger set of tables? I would assume that the static approach grows linear with the number of tables, as each new table leads to another class being loaded, where the dump approach stays mostly the same, as adding another new table just means to read two more lines from a file.

### CODE REVIEW

**Comment by mringler** [comment]
Psalm fails in an unchanged class, I don't understand the error:
```
ERROR: InvalidClass - src/Propel/Generator/Manager/AbstractManager.php:326:32 - Class, interface or enum XsltProcessor has wrong casing (see https://psalm.dev/007)
                    $xsl = new XsltProcessor();

```

File: src/Propel/Runtime/Map/DatabaseMap.php
Comment (dereuromark) [inline src/Propel/Runtime/Map/DatabaseMap.php:140]:
```suggestion
     * @return array<string, mixed> A dump that can be loaded again with {@link DatabaseMap::loadMapsFromDump()}
```
or even more

File: src/Propel/Runtime/ServiceContainer/StandardServiceContainer.php
Comment (dereuromark) [inline src/Propel/Runtime/ServiceContainer/StandardServiceContainer.php:274]: same

File: src/Propel/Runtime/Map/DatabaseMap.php
Comment (mringler) [reply src/Propel/Runtime/Map/DatabaseMap.php]: Hey Mark, thank you for the review. Played around with psalm a bit, let me know what you think

File: src/Propel/Runtime/Map/DatabaseMap.php
Comment (dereuromark) [inline src/Propel/Runtime/Map/DatabaseMap.php:145]: This looks like array<mixed, ..> instead of a object return though.

Comment (mringler) [reply src/Propel/Runtime/Map/DatabaseMap.php:145]: Yes, good that you mention it, I am not sure how much I like that. Apparently psalm allows type aliases, which `\Propel\Runtime\Map\TableMapDump` is, I put the declaration in the docblock of the class:
```php
* @psalm-type \Propel\Runtime\Map\MapType = 'tablesByName' | 'tablesByPhpName'
 * @psalm-type \Propel\Runtime\Map\TableMapDump array<\Propel\Runtime\Map\MapType, array<string, class-string<\Propel\Runtime\Map\TableMap>>>
 ```
On one hand, it is nice, because we can make sure that the load and dump methods always use the exact same type, and through an import with `@psalm-import-type`, it can be used in the `StandardServiceContainer` too. So I think in general this is the "correct" approach.
On the other hand, a lot about it seems confusing. I had to give the alias a name that looks like a namespaced type (otherwise it wouldn't work with phpcs, which changed "TableMapDump" to "\Propel\Runtime\Map\TableMapDump", causing a no such type error). Also IDE support for type aliases is not (yet) what it could be, at least in VS Code and Eclipse. Overall, I have the impression that type aliases as a concept is still rather new in PHP, which may or may not keep us from using it. What do you think?

Comment (dereuromark) [reply src/Propel/Runtime/Map/DatabaseMap.php:145]: I think the main docblock (return) should reflect trivial reality (array)
The on top specific phpstan and psalm tags can contain the "new" concepts.

Comment (mringler) [reply src/Propel/Runtime/Map/DatabaseMap.php:145]: Like this then?
```php
    /**
     * Dump table maps. Used during configuration generation.
     *
     * @psalm-return \Propel\Runtime\Map\TableMapDump
     *
     * @return array A dump that can be loaded again with {@link DatabaseMap::loadMapsFromDump()}
     */
```

**Comment by profuel** [comment]
I like the idea, it looks like a next step to optimize loading of a tables.
I cannot promise fast test, but in a week - I will provide you with results.

**Comment by profuel** [comment]
I've tested this locally.
It has 0.5-2 ms `wasted` time of every request which is simply awesome, comparing to the original 10+ ms.

So I'm fully for this solution.

BTW, \Propel\Runtime\ServiceContainer\StandardServiceContainer and especially setConnectionManager has a major change, which lead to a minor change on our project side :)

File: tests/Propel/Tests/Runtime/Map/DatabaseMapTest.php
Comment (dereuromark) [inline tests/Propel/Tests/Runtime/Map/DatabaseMapTest.php:131]:
```suggestion
    protected function assertTableMapContains(DatabaseMapWithGetters $databaseMap, $tableName, $phpTableName, $tableMapClass)
    {
```

I wonder if CS is not active for tests.

**Comment by dereuromark** [comment]
@profuel I dont see any changes to setConnectionManager() etc, can you clarify the major changes that have implications for you?

**Comment by profuel** [comment]
@dereuromark my fault, we have an older Propel on the project. All good here.

**Review by dereuromark — APPROVED** [review]
Given the analyis/QA I will now approve this
With a 2nd approve we have green light for merge.

File: tests/Propel/Tests/Runtime/Map/DatabaseMapTest.php
Comment (mringler) [reply tests/Propel/Tests/Runtime/Map/DatabaseMapTest.php]: Yes, tests are currently not linted

**Comment by mringler** [comment]
Thank you @profuel for the tests!

**Comment by profuel** [comment]
@mringler my pleasure. We need this change on the project, so - mutual goal :)

**Comment by dereuromark** [comment]
Any other approvers? I am only waiting for this for the PR to get merged.

**Review by yaroslav-spryker — APPROVED** [review]

**Pull request merged**
